### PR TITLE
[SS-54] Add dialog to allow local deletion if network deletion fails

### PR DESCRIPTION
**Pull request by AL-Session**
### Contributor checklist

- [X] I have tested my contribution on these devices:
 * Virtual Pixel 3a, Android 9 API 28
 * Virtual Pixel 3a, Android 14 API 34
- [X] My contribution is fully baked and ready to be merged as is
- [X] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description
New functionality: If the user goes to **Settings > Clear Data** and chooses to "Clear device and network", then should the network data deletion fail the user is informed of this issue and asked if they would like to delete local device data only instead. If they proceed then local device data is deleted as if that had been their first choice.

One way to quickly test this functionality & see this new modal dialog is to:
- Get to the point where you can "Clear device and network", and then
- Drag down from the top of the screen to show the Android settings and turn **Flight Mode** on, then 
- Click the current modal's "Clear" button.

As we're in flight mode, a short while after we attempt to perform the network deletion it will fail and the new modal displaying "_An unknown error occurred and your data was not deleted. Do you want to delete your data from just this device instead?_ [Clear Device] [Cancel]". At this point the [Clear Device] button will do just that, while the [Cancel] button will cancel out of the operation.

### CODE REVIEW

**Comment by AL-Session** [comment]
Reversion of `dev` merge & `libsession-util` addition has been undone following dev process meeting.

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 1]
```diff
@@ -86,8 +99,12 @@ class ClearAllDataDialog : DialogFragment() {
                     binding.dialogDescriptionText.setText(R.string.dialog_clear_all_data_clear_device_and_network_confirmation)
                 }
                 Steps.DELETING -> { /* do nothing intentionally */ }
+                Steps.RETRY_LOCAL_DELETE_ONLY_PROMPT -> {
+                    binding.dialogDescriptionText.setText(R.string.clearDataErrorDescriptionGeneric)
+                    binding.clearAllDataButton.text = getString(R.string.clearDevice)
+                }
             }
-            binding.recyclerView.isGone = step == Steps.NETWORK_PROMPT
+            binding.recyclerView.isGone = step == Steps.NETWORK_PROMPT || step == Steps.RETRY_LOCAL_DELETE_ONLY_PROMPT
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:107]: this was probably only `isGone` because there was only one step to check, now there's 2, it's simpler to say `isVisible = either of the other two`, which is kinda better anyway, because we're checking the 2 steps that have items, not that it's not in a step that doesn't have items... yeah. Better still would be to extract the logic out of the view, but could be overkill here.

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 2]
```diff
@@ -24,17 +27,26 @@ import org.thoughtcrime.securesms.dependencies.DatabaseComponent
 import org.thoughtcrime.securesms.util.ConfigurationMessageUtilities
 
 class ClearAllDataDialog : DialogFragment() {
+    private val TAG = "ClearAllDataDialog"
+
     private lateinit var binding: DialogClearAllDataBinding
 
     enum class Steps {
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:34]: `private enum class Step` ?

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 3]
```diff
@@ -97,46 +114,50 @@ class ClearAllDataDialog : DialogFragment() {
...
-            if (!deleteNetworkMessages) {
                 try {
                     ConfigurationMessageUtilities.forceSyncConfigurationNowIfNeeded(requireContext()).get()
                 } catch (e: Exception) {
-                    Log.e("Loki", "Failed to force sync", e)
+                    Log.e(TAG, "Failed to force sync when deleting data", e)
+                    Handler(Looper.getMainLooper()).post {
+                        Toast.makeText(ApplicationContext.getInstance(requireContext()), R.string.errorUnknown, Toast.LENGTH_LONG).show()
+                    }
+                    return@launch
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:131]: I think when we have to `return@` we should consider making this a function, as these returns are always hard to follow.

A well named function is self-dcoumenting code.

File: app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java [hunk 4]
```diff
@@ -501,7 +501,8 @@ private void loadEmojiSearchIndexIfNeeded() {
         });
     }
 
-    public void clearAllData(boolean isMigratingToV2KeyPair) {
+    // Method to clear the local data - returns true on success otherwise false
```
[the remark this answers is not in the document]
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java:504]: Do you mean like this?
```
/**
 * Clear all local profile data and message history then restart the app after a brief delay.
 * @param isMigratingToV2KeyPair whether we're upgrading to a more recent V2 key pair or not.
 * @return true on success, false otherwise.
 */
```

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 5]
```diff
@@ -97,46 +114,50 @@ class ClearAllDataDialog : DialogFragment() {
...
+                if (!success) {
+                    Handler(Looper.getMainLooper()).post {
+                        Toast.makeText(ApplicationContext.getInstance(requireContext()), R.string.errorUnknown, Toast.LENGTH_LONG).show()
+                    }
                 }
-            } else {
-                // finish
-                val result = try {
+            }
+            else if (deletionScope == DeletionScope.DeleteBothLocalAndNetworkData) {
```
[the remark this answers is not in the document]
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:141]: Done

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt (hunk 2 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:34]: Done

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 6]
```diff
@@ -97,46 +114,50 @@ class ClearAllDataDialog : DialogFragment() {
...
+                    }
+                    return@launch
                 }
-                ApplicationContext.getInstance(context).clearAllData(false)
-                withContext(Dispatchers.Main) {
-                    dismiss()
+                val success = ApplicationContext.getInstance(context).clearAllData(false)
+                withContext(Dispatchers.Main) { dismiss() }
+                if (!success) {
+                    Handler(Looper.getMainLooper()).post {
```
[the remark this answers is not in the document]
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:136]: Done

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt (hunk 1 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:107]: Flipped logic to use `isVisible` rather than `isGone`

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt (hunk 3 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:131]: Separated out local delete steps into a separate method.

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 7]
```diff
@@ -97,46 +115,52 @@ class ClearAllDataDialog : DialogFragment() {
...
-                withContext(Dispatchers.Main) {
-                    dismiss()
-                }
-            } else {
-                // finish
-                val result = try {
+    private fun clearAllData(deletionScope: DeletionScope) {
+
+        clearJob = lifecycleScope.launch(Dispatchers.IO) {
+            withContext(Main) { step = Steps.DELETING }
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:140]: may as well remove the scheduling and put this on Line 138, do it synchronously is safer, no?

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 8]
```diff
@@ -97,46 +115,52 @@ class ClearAllDataDialog : DialogFragment() {
...
+        try {
+            ConfigurationMessageUtilities.forceSyncConfigurationNowIfNeeded(requireContext()).get()
+        } catch (e: Exception) {
+            Log.e(TAG, "Failed to force sync when deleting data", e)
+            withContext(Main) {
+                Toast.makeText(ApplicationContext.getInstance(requireContext()), R.string.errorUnknown, Toast.LENGTH_LONG).show()
+            }
+            return
+        }
+        val success = ApplicationContext.getInstance(context).clearAllData(false)
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:128]: Maybe more readable, saving the val, and the double scheduling.

```
ApplicationContext.getInstance(context).clearAllData(false).let { success ->
    withContext(Main) {
        dismiss()
        if (!success) {
```

File: app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt [hunk 9]
```diff
@@ -97,46 +115,52 @@ class ClearAllDataDialog : DialogFragment() {
...
-                }
-            } else {
-                // finish
-                val result = try {
+    private fun clearAllData(deletionScope: DeletionScope) {
+
+        clearJob = lifecycleScope.launch(Dispatchers.IO) {
+            withContext(Main) { step = Steps.DELETING }
+
+            if (deletionScope == DeletionScope.DeleteLocalDataOnly) {
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/preferences/ClearAllDataDialog.kt:142]: `when` would be nice here.

**Review by bemusementpark — APPROVED** [review]
looks good!
